src/rgw/rgw_datalog.h

```diff
@@ -50,23 +50,29 @@ struct rgw_data_change {
   DataLogEntityType entity_type;
   std::string key;
   ceph::real_time timestamp;
+  uint64_t gen_id;
 
   void encode(ceph::buffer::list& bl) const {
-    ENCODE_START(1, 1, bl);
+    ENCODE_START(2, 2, bl);
```

Comment:
For rolling upgrades, we should change this so it sends the 1,1 version (that doesn't know about gen ids and can be decoded by old versions) when we're in generation zero. That way we only break compatibility when we reshard.

Reply:
if an upgraded radosgw encodes this with v2, a non-upgraded radosgw will still be able to decode it as v1 because the `compat` version allows it. the `DECODE_FINISH()` macro will just skip past the unrecognized bytes at the end

conversely, it could be helpful to raise an error in the case where a datalog entry with gen>0 would be misinterpreted by older clients. how about this?
```c++
const uint8_t compat = (gen == 0) ? 1 : 2; // require decoders to recognize v2 when gen>0
ENCODE_START(2, compat, bl);
```

Reply:
That sounds fine to me!

Reply:
we only try to decode these entries in `RGWDataChangesBE::list()`, so an error there would effectively prevent old radosgws from serving those entries to other zones in `RGWOp_DATALog_List`